### ISSUE TITLE
How to deal with incompatibilities between conda-forge and r channel

### ISSUE THREAD

**Issue by johanneskoester** [bioconda/bioconda-recipes#2921]
@bioconda/core, recently, we had numerous problems with packages from e.g. the R channel pulling dependencies from conda-forge. I described the issue in conda/conda#3721.

### Current workaround
So far, our workaround for the issue is to fix channel priorities such that conda-forge has least priority.
This fixes the issue only for packages from the default and the R channel though. For the conda-forge channel, this actually should rather make it worse, because dependencies of those packages might be pulled from R and defaults if they don't happen to be pinned to a newer version.
Another issue is that the way channel priorities are configured in conda is highly error-prone and unintuitive. The order in condarc decides, there is a semantic difference between add and append, ...
This is illustrated by the fact that we needed 3 tries to get the priorities right in the last 4 weeks or so.

### Status of possible fix in conda or conda-build
As you will see in conda/conda#3721, response from the involved people has been quite limited. There is one possible solution (pinning of build args in addition to versions, conda/conda-build#1142). However, this solution seems to be quite far away, and all my alternative suggestions have not found triggered any actions so far.

### Solutions
I think we should at least think about how we can get control again. Hence, I would like to discuss the following approach:

#### Solution 1
* we remove conda-forge from the list of used channels
* we identify packages we need from conda-forge
* we put the corresponding feedstocks as git submodules into e.g. a subfolder of this repo. Of course conda-forge will get the credit.
* we build those recipes with our build system, having only defaults and R channel activated
* we try to keep the list of included conda-forge packages mutually exclusive with defaults and R. This way, no binary incompatibility issues can occur

If we don't manage to maintain the mutual exclusivity (e.g. because we don't manage to propagate fixes to defaults or R), we could even use an analog strategy for the R channel. But I think it should be doable without that.

The downside of this strategy is that we need some additional build time and space for those conda-forge packages that we include.

#### Solution 2
Move the entire R channel into conda-forge. The R channel consists of ~230 packages. It will be quite some effort, and we should try to be quick with this in order to minimize inconvenience for our users.

#### Solution 3
Rebuild R channel with conda-forge as dependency.

EDIT: Added solution 2, suggested by @bgruening.

What do you think?

**Comment by bgruening** [comment]
I don't see it as dark as your post sounds like. anaconda tries imho to find a way to rebuild their collection of packages and make it more open with conda-forge. This process is not finished and things was getting more complicated with the conda-build 2 release and the short-prefix problems. What is a little bit sad is that there is a lack of communication and the future of `r`, `anaconda` and `conda-forge` and how they relate is not entirely clear - at least for me.

> * we remove conda-forge from the list of used channels

Not an option for me. I want to have communities working together and not against each other.

> * we identify packages we need from conda-forge

I think this is hard to do, every contributor has different needs, so we can not do this in advance.

> * we put the corresponding feedstocks as git submodules into e.g. a subfolder of this repo.
> * we build those recipes with our build system, having only defaults and R channel activated
> * we try to keep the list of included conda-forge packages mutually exclusive with defaults and R. This way, no binary incompatibility issues can occur

I'm really afraid of these doubled dependencies and our space requirements. With our current growth we will reach the 100GB limit in a year - new/copied packages will increase this even more. We should move packages to conda-forge and not vice-versa.

To find a solution for this problem in general it would be nice if we can talk with @msarahan and conda-forge (ping @jakirkham). My gut feeling is we should help them migrating packages from `r` to `conda-forge` and getting rid of the `r` channel.

**Comment by johanneskoester** [comment]
@bgruening good to hear that it is not as bad as my impression was.

>> we remove conda-forge from the list of used channels

> Not an option for me. I want to have communities working together and not against each other.

To clarify: this was not about working against conda-forge. On the contrary, I still want to contribute whenever possible. This is only about a different way of including a channel. We still use conda-forge as a dependency and contribute back. The only difference is that rebuilds of some of their recipes are also available directly in bioconda. Still all credit goes back to conda-forge. And since we only use submodules and no fork, there is also no duplication of work. But you might be right that there are better options.

> To find a solution for this problem in general it would be nice if we can talk with @msarahan and conda-forge (ping @jakirkham). My gut feeling is we should help them migrating packages from r to conda-forge and getting rid of the r channel.

In conda/conda#3721 I was referred to the new pinning system as the solution. But that is a huge task, and quite far away. I think we need a more short term solution and porting the R channel into conda-forge does not sound particularly short term either. But maybe I am wrong. If we work together on that now, I think it would be preferable indeed. R contains about 230 recipes I think. I would be happy to help, but we should really do it now then.

**Comment by bgruening** [comment]
@johanneskoester I understand too less about conda-smithy but I think it is a more or less automatic system to create the feedstocks - so I think technically it might be "easy". But not sure about the politics behind it. Let's join the next conda-forge meeting and see if @msarahan @jakirkham can help us here with input.

**Comment by saketkc** [comment]
I recently bumped into an issue, I am not sure if it has been considered before:

What is the best way to avoid package name conflicts? For example, there is Piranha, a CLIP/RIP-Seq peak caller in [bioconda](https://github.com/bioconda/bioconda-recipes/pull/1711) and Piranha, a computer algebra system in [conda-forge](https://github.com/conda-forge/staged-recipes/pull/1918)

While updating my packages, it makes a spurious update to Piranha(So I have a computer algebra system instead of a Peak caller):

```
The following packages will be SUPERCEDED by a higher-priority channel:

    piranha:         1.2.1-gsl1.16_2    bioconda    --> 0.5-0                              conda-forge
```

**Comment by johanneskoester** [comment]
@saketkc this should be handled by the channel priorities. Can you please post the output of  `conda config --get`?

**Comment by saketkc** [comment]
``` 
$ conda config --get
--add channels 'defaults'   # lowest priority
--add channels 'r'
--add channels 'bioconda'
--add channels 'conda-forge'   # highest priority
```

Shouldn't `conda-forge` always supersede `bioconda` (as it now does)

**Comment by johanneskoester** [comment]
Ideally, this would depend on your use case. Currently, we need the channel priorities to avoid the incompatibilities outlined above. The correct channel order is listed on bioconda.github.io. I am quite sure that it will fix your problem.

**Comment by jakirkham** [comment]
Also you guys may want to weigh in on our meeting schedule. We are changing it because Phil is now in Australia. So we need to find some new times that work. It may be touch and go for a few weeks while we iron out what works and plan around holidays.

xref: http://doodle.com/poll/pgaw6c654wenigmy

**Comment by johanneskoester** [comment]
Thanks @jakirkham.
So, to be concrete again: @mingwandroid, what is your opinion on duplicating the R channel in conda-forge. Do we have your permission to use the recipes in conda/conda-recipes as a base?

**Comment by mingwandroid** [comment]
@johanneskoester, Yes, I currently handle all of the R packages.

I would rather wait to see what happens with the new build system that @msarahan is working on at present. I expect to adopt a more `conda-forge`-like layout for the R packages once that is ready. At that point, if you still feel that copying R into `conda-forge` is the way to go then it should be easier for you to do.

A lot of this will be repeats of what's already been said so apologies in advance but as @bgruening said:

> What is a little bit sad is that there is a lack of communication and the future of r, anaconda and conda-forge and how they relate is not entirely clear - at least for me

**Solution 1**
This doesn't really fix anything AFAICT, you are just moving incompatible recipes from one place to another. The mutual exclusivity mentioned cannot exist otherwise you would not have a problem to speak of. The problem is the incompatibility (or explicitly the fact that the packages are different at the C/C++ (and other) interface level).

**Solution 2**
You have permission to take the recipes and add them all to `conda-forge`. I believe recipes for building software are not copyrightable (beyond any significantly large patches or bit of logic) in the same way that those for making cakes are not copyrightable; even if they are copyrightable I am pretty sure Continuum would not object, @csoja I would appreciate confirmation on that point though.

Do you need to do this at present? Given the ordering in https://github.com/bioconda/bioconda-recipes/issues/2921#issuecomment-260590358 do you still have a definite problem or are you trying to head-off likely future problems? Adding the same packages to both `defaults` and `conda-forge` is (of course) the event that creates binary incompatibility and should only be done after the most serious consideration and after exploring *all* other options. It usually prevents people who need or want to use both channels from doing so, so I guess implicit in this solution is dropping `defaults` altogether which is perhaps an option for `bioconda`, but it will probably not be for all people who want to use `bioconda`.

**Solution 3**
This would prevent using the `r` and `defaults` channels together. For Continuum that is not going to work as things currently stand.

**My take on it**
If you have compatibility issues between the R infrastructure and `conda-forge` then there are only a handful of C/C++ libraries that need to be brought back to binary compatibility and it is my opinion that this is (currently) the best way forward. It would need some compromise for sure and someone to drive it. From the Continuum/`defaults` side I would try my best to make this work.

**Comment by johanneskoester** [comment]
@csoja, may I ask for your confirmation again? Thanks.

**Comment by csoja** [comment]
Thanks for pinging me again - I had missed this.  Yes, you are welcome to use the public recipes to add the R packages to conda-forge.  It would be great if you included the comment noting the original recipe came from Continuum Analytics.

**Comment by johanneskoester** [comment]
Thanks! Of course we will give the credit to you.

**Comment by bgruening** [comment]
@johanneskoester are you planning to work on it? I guess we need to adopt every package slightly and create a PR for every single package...?

Thanks @csoja
Thanks

**Comment by johanneskoester** [comment]
Yes. I have started with java-jdk (a build dependency of R). I plan to do one or two packages per day. Might be that I first have to update the CRAN parser in conda, such that it outputs the conda-forge specific format.

**Comment by mingwandroid** [comment]
Are you porting openjdk? This is something I'm interested in. If I can help let me know.

**Comment by bgruening** [comment]
@johanneskoester if you can create a tracking issue I will try to help you as much as I can.

**Comment by jakirkham** [comment]
Agreed OpenJDK would be a nice add.

I think there is an Oracle JDK repackaging PR at staged-recipes. Not sure what state it is in.

Also feel free to create an issue at staged-recipes with a listing of things you need/plan to package as part of the R effort. Others may end up chipping in. So it is good to stay organized.

**Comment by johanneskoester** [comment]
@mingwandroid, @jakirkham I am currently porting the Zulu OpenJDK build that we already had here in Bioconda, see PR conda-forge/staged-recipes#2002. For the tracking issue, see conda-forge/staged-recipes#2009. I would really appreciate any help. Especially once I have managed to port the prereqs and start porting CRAN packages.

**Comment by mingwandroid** [comment]
Ah, we have `Zulu OpenJDK` on defaults already. They don't give the 32-bit builds out for free unfortunately so that causes problems. For `r-base` I've been just using the 64-bit one in the 32-bit build, but I'm having to avoid `conda-build` for that and just setting PATH before running `conda-build`. Compiling `OpenJDK` from source would give a fully integrated solution here but I understand it's not trivial.

**Comment by johanneskoester** [comment]
@mingwandroid cool, I was not aware of that. That is sufficient for me. I guess I'll close my JDK PR on staged-recipes then.

**Comment by mingwandroid** [comment]
Well, I see no reason to not add OpenJDK to `conda-forge`. Binary repackaging the same thing should end up with the same thing so there should be no compatibility issues to worry about.

**Comment by johanneskoester** [comment]
The only reason could be that old versions might be needed in bioconda that are not provided in the anaconda channel. @acaprez you added Java 6, do we still need that? Also @chapmanb I think you once mentioned that GATK needs Java 7?

**Comment by chapmanb** [comment]
GATK 3.6+ and MuTect2 now use Java 8 so we use that as the default. Older GATK and MuTect are Java 7 only so it would be useful to have it available as well. We've been using the Zulu OpenJDK in bioconda without any issues.

**Comment by acaprez** [comment]
@johanneskoester - I added Java 6 for `picard` version 1.97.  I don't recall the exact details now, but it was unhappy using a newer Java version, which is why I added 6.
